### PR TITLE
[kube] use local image for kubernetes integration tests

### LINKED ISSUE #45986: `TestKube` flakiness
#### Link(s) to logs

- https://github.com/gravitational/teleport/actions/runs/10603785318/job/29388916169

#### Relevant snippet

```
=== RUN   TestKube
    kube_integration_test.go:149: 
        	Error Trace:	/__w/teleport/teleport/integration/kube_integration_test.go:149
        	            				/__w/teleport/teleport/integration/kube_integration_test.go:171
        	Error:      	Condition never satisfied
        	Test:       	TestKube
--- FAIL: TestKube (61.77s)
FAIL integration.TestKube (61.77s)
FAIL integration

=== Failed
=== FAIL: integration TestKube (61.77s)
    kube_integration_test.go:149: 
        	Error Trace:	/__w/teleport/teleport/integration/kube_integration_test.go:149
        	            				/__w/teleport/teleport/integration/kube_integration_test.go:171
        	Error:      	Condition never satisfied
        	Test:       	TestKube
```

### PR DESCRIPTION
**Pull request by tigrato**
This PR replaces the use of the `nginx:latest` Docker image in Kubernetes integration tests with a custom-built image based on `alpine:3.20.3`. This custom image includes a shell for `kubectl exec` integration tests and a compiled binary to act as an HTTP server.

This change addresses recent issues where test workflows failed due to problems downloading the `nginx` image.

### CODE REVIEW

File: fixtures/alpine/webserver
Comment (r0mant) [inline fixtures/alpine/webserver:1]: Do we need to commit this binary? It looks like it's being go-built in the workflow itself.

File: fixtures/alpine/alpine-3.23.amd64.tar
Comment (codingllama) [inline fixtures/alpine/alpine-3.23.amd64.tar:1]: What is the patch version? Could we get a checksum to go with it too?

File: fixtures/alpine/webserver
Comment (codingllama) [reply fixtures/alpine/webserver:1]: +1 for no committed binaries

File: integration/kube_integration_test.go
Comment (codingllama) [inline integration/kube_integration_test.go:1931]: Please add a comment here explaining where the image comes from. Maybe even link the workflow so folks know how to reproduce it locally.

File: fixtures/alpine/alpine-3.23.amd64.tar
Comment (codingllama) [reply fixtures/alpine/alpine-3.23.amd64.tar:1]: Mentioning the source of the image would be nice too.

Comment (r0mant) [reply fixtures/alpine/alpine-3.23.amd64.tar:1]: There's also discrepancy between versions: this file name is `3.23` but the base Docker image is `3.20.3`. Typo in this file name?

File: fixtures/alpine/Dockerfile
Comment (fheinecke) [inline fixtures/alpine/Dockerfile:5]:
```suggestion
CMD [ "/webserver" ]

```

File: fixtures/alpine/alpine-3.23.amd64.tar
Comment (fheinecke) [reply fixtures/alpine/alpine-3.23.amd64.tar:1]: IMO legal should review this. Including an entire container image worth of OS libraries/tools/external software means that we're now distributing a lot more packages.

**Review by fheinecke — COMMENTED** [review]
Rather than doing all this, can we configure the workflow to use a pull-through cache in a private ECR registry? Then we don't have to worry about licenses (we won't be redistributing anything), we won't be adding lots of binaries to the repo, and it'll be easier for us to keep up to date.

File: fixtures/alpine/webserver
Comment (tigrato) [reply fixtures/alpine/webserver:1]: it was a typo

File: fixtures/alpine/alpine-3.20.3.amd64.sum
Comment (codingllama) [inline fixtures/alpine/alpine-3.20.3.amd64.sum:1]: Could you refer to a hash and image name that can be recognized from https://dl-cdn.alpinelinux.org/alpine/v3.20/releases/x86_64/ (or whatever source you used)?

File: fixtures/alpine/alpine-3.23.amd64.tar
Comment (tigrato) [reply fixtures/alpine/alpine-3.23.amd64.tar:1]: >IMO legal should review this. Including an entire container image worth of OS libraries/tools/external software means that we're now distributing a lot more packages.

Shouldn't this apply even if we vendored the image in gha container repo?

File: fixtures/alpine/alpine-3.20.3.amd64.sum
Comment (tigrato) [reply fixtures/alpine/alpine-3.20.3.amd64.sum:1]: this is a simple docker pull && docker save. We can run `docker trust inspect alpine:3.20.3` to inspect the signature of the image but I am just trying to unblock everyone.

Comment (codingllama) [reply fixtures/alpine/alpine-3.20.3.amd64.sum:1]: Unblocking folks is important, but we are adding a lump of binaries to the repo so the provenance of those must be clear. Not just today, but also in the future when/if someone comes along and puts a newer .tar.gz in there.

> We can run docker trust inspect alpine:3.20.3

That's not a bad idea. That would detect a modified image, yes?

File: fixtures/alpine/README.md
Comment (codingllama) [inline fixtures/alpine/README.md:12]: FWIW I don't get the same .tar as you when I do that. I also suspect this would save an arm64 blob if I ran this command as-is.

File: fixtures/alpine/alpine-3.20.3.amd64.sum
Comment (tigrato) [reply fixtures/alpine/alpine-3.20.3.amd64.sum:1]: That would detect if the image is signed by the expected key.

File: fixtures/alpine/README.md
Comment (tigrato) [reply fixtures/alpine/README.md:12]: Yes. It saves your default platform. You must pass --platform linux/amd64

Comment (codingllama) [reply fixtures/alpine/README.md:12]:
```shell
$ docker pull --platform linux/amd64 docker.io/library/alpine:3.20.3
> 3.20.3: Pulling from library/alpine
> Digest: sha256:beefdbd8a1da6d2915566fde36db9db0b524eb737fc57cd1367effd16dc0d06d
> Status: Image is up to date for alpine:3.20.3
> docker.io/library/alpine:3.20.3
> 
> What's next:
>     View a summary of image vulnerabilities and recommendations → docker scout quickview docker.io/library/alpine:3.20.3

# Note: docker save doesn't take the --platform flag
$ docker save -o alpine-3.20.3.amd64.tar docker.io/library/alpine:3.20.3
$ shasum -a256 alpine-3.20.3.amd64.tar 
> c8f6acdead29241e5034e5382d3df3902919b8c5edcd24c8b385ba38a3f7759c  alpine-3.20.3.amd64.tar
```

This seems to give me a different hash every time, so it's hard to verify like that. (The image digest is deterministic though.)

Comment (tigrato) [reply fixtures/alpine/README.md:12]: IIRC, docker images don't follow any particular order when saving layers to disk

File: fixtures/alpine/alpine-3.23.amd64.tar
Comment (fheinecke) [reply fixtures/alpine/alpine-3.23.amd64.tar:1]: Not if we used private ECR, then we're only distributing it within the company

Comment (tigrato) [reply fixtures/alpine/alpine-3.23.amd64.tar:1]: We base only our test images for the Kube integration tests.

Is there a way of storing this image in any other private repo and make oss download it?

File: fixtures/alpine/README.md
Comment (tigrato) [reply fixtures/alpine/README.md]: On [d99891f](https://github.com/gravitational/teleport/pull/48321/commits/d99891fb34952a4854aad33f918c4646ca6f9861), I sourced the files from alpine cdn and added gpg validation.

File: fixtures/alpine/alpine-3.23.amd64.tar
Comment (tigrato) [reply fixtures/alpine/alpine-3.23.amd64.tar:1]: On [d99891f](https://github.com/gravitational/teleport/pull/48321/commits/d99891fb34952a4854aad33f918c4646ca6f9861), I sourced the files from alpine cdn and added gpg validation.

File: fixtures/alpine/README.md
Comment (codingllama) [inline fixtures/alpine/README.md:9]: > To avoid distributing Alpine Linux within the Teleport repository, we download it fresh from the Alpine CDN with each build

Doesn't that defeat the purpose of the PR?

**Review by codingllama — APPROVED** [review]

File: fixtures/alpine/README.md
Comment (tigrato) [reply fixtures/alpine/README.md:9]: Yes, but given the distribution concerns that @fheinecke mentioned and the greater reliability of the Alpine CDN compared to Docker Hub, I hope we can proceed with this approach.

Ideally, I would prefer to keep the Alpine image within the GitHub repository itself, though this could potentially raise some legal concerns. We can always move to any private repo at any time

Comment (codingllama) [reply fixtures/alpine/README.md:9]: +1 for Roman's suggestions.

Comment (zmb3) [reply fixtures/alpine/README.md:9]: Or just manually push it to private ECR and update the workflow to try ECR first and fall back to Docker Hub?

This repo is super bloated - it would be nice to avoid committing additional large binary blobs.

Comment (tigrato) [reply fixtures/alpine/README.md:9]: @r0mant https://github.com/gravitational/teleport/pull/48321#discussion_r1826310685

I have no idea if the concerns are valid or not. 

@zmb3 Teleport OSS doesn’t have access permissions for our private ECR. We’re using GitHub’s container registry instead to store the buildbox images. We can do the same here but it's probably worth having someone else working on it.

File: fixtures/alpine/README.md
Comment (codingllama) [inline fixtures/alpine/README.md:26]:
```suggestion
$ curl -fSsLO https://dl-cdn.alpinelinux.org/alpine/v3.20/releases/x86_64/alpine-minirootfs-$ALPINE_VERSION-x86_64.tar.gz
$ curl -fSsLO https://dl-cdn.alpinelinux.org/alpine/v3.20/releases/x86_64/alpine-minirootfs-$ALPINE_VERSION-x86_64.tar.gz.asc
$ curl -fSsLO https://dl-cdn.alpinelinux.org/alpine/v3.20/releases/x86_64/alpine-minirootfs-$ALPINE_VERSION-x86_64.tar.gz.sha256
```

File: fixtures/alpine/README.md
Comment (hugoShaka) [reply fixtures/alpine/README.md]: > Doesn't that defeat the purpose of the PR?

Downloading before and loading manually in kind will very likely solve the hole issue, the issue can be caused by two things:
- docker hub rate-limits. Since 2022, docker hub enforces agressive pull limits per IP address on unauthenticated requests. Downloading artifacts from a more generous CDN will very likely solve the issue.
- docker in docker networking. Docker in docker setups typically have broken networks caused by MTU mismatch and several NAT layers blocking control frames top reach the caller back. This is a typical dind problem, the dind container has a 1500 MTU while the underlying runnner network has 1460 because of the CNI encapsulation. Downloading the external assets outside of the dind will very likely solve the whole issue and we'll go back to 100% success rate.

Also I agree with @zmb3  and we should try to avoid committing more bloat to the repo.

File: fixtures/alpine/README.md
Comment (codingllama) [inline fixtures/alpine/README.md:22]:
```suggestion
```bash
$ export ALPINE_VERSION=3.20.3

```

**Review by codingllama — APPROVED** [review]

**Review by hugoShaka — APPROVED** [review]

**Review by r0mant — APPROVED** [review]

**Review by codingllama — APPROVED** [review]

File: fixtures/alpine/README.md
Comment (fheinecke) [reply fixtures/alpine/README.md]: > What are the distribution concerns? It is not a part of the product, we are not modifying it in any way and are not going to distribute it with our release assets. 

Just because it's not a part of the product does not mean that we're not distributing it. This is a ton of software that we don't hold the copyright to, in a publicly-accessible (and indexed) file sharing service. As I see it, the only difference between uploading the image here, and uploading something like Microsoft Windows, Adobe Photoshop, movies, songs, etc. is that we _might_ have a legal right to do so, due to the licenses that each piece of software grants us. There are likely other stipulations to these licenses as well, such as making the source code available as easily as the binaries themselves.

Whether or not we can or should legally do this is a determination that somebody trained in copyright law should make.

**Pull request merged**
